figtree.go

```diff
@@ -691,375 +707,1044 @@ func (m *Merger) mustIgnore(name string) bool {
+					dest.Set(reflectedSrc)
+				} else {
+					// slices are mutable, so create a brand new shiny one
+					cp := reflect.MakeSlice(reflectedSrc.Type(), reflectedSrc.Len(), reflectedSrc.Len())
+					reflect.Copy(cp, reflectedSrc)
+					dest.Set(cp)
+				}
+			default:
+				dest.Set(reflectedSrc)
 			}
-			if destOptionValue.Kind() == reflect.String && src.Kind() != reflect.String {
-				option.SetValue(fmt.Sprintf("%v", src.Interface()))
-				option.SetSource(m.sourceFile)
-				Log.Debugf("assignValue: assigned %#v to %#v", destOptionValue, src)
-				return
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if dest.Kind() == reflect.Bool && reflectedSrc.Kind() == reflect.String {
+		b, err := strconv.ParseBool(reflectedSrc.Interface().(string))
+		if err != nil {
+			return false, errors.Wrapf(err, "%s is not assignable to %s, invalid bool value %#v", reflectedSrc.Type(), dest.Type(), reflectedSrc)
+		}
+		dest.Set(reflect.ValueOf(b))
+		return true, nil
+	}
+
+	if dest.Kind() == reflect.String && reflectedSrc.Kind() != reflect.String && stringType.AssignableTo(dest.Type()) {
+		switch reflectedSrc.Kind() {
+		case reflect.Array, reflect.Slice, reflect.Map:
+			return false, errors.WithStack(
+				notAssignableError{
+					srcType:        reflectedSrc.Type(),
+					dstType:        dest.Type(),
+					sourceLocation: NewSource(m.sourceFile, WithLocation(coord)),
+				},
+			)
+		case reflect.Struct:
+			// we generally dont want to assign structs to a string
+			// unless that struct is an option struct in which case
+			// we use convert the value
+			if option := toOption(reflectedSrc); option != nil {
+				dest.Set(reflect.ValueOf(fmt.Sprintf("%v", option.GetValue())))
+			}
+			return false, errors.WithStack(
+				notAssignableError{
+					srcType:        reflectedSrc.Type(),
+					dstType:        dest.Type(),
+					sourceLocation: NewSource(m.sourceFile, WithLocation(coord)),
+				},
+			)
+		default:
+			// if we have a scalar node we want to convert to a string, just use
+			// the literal value tokenized from the document, this will
+			// allow values like `False` to be preserved as a case-sensitive
+			// string rather than being converted to a bool, the back to a string.
+			if src.node != nil && src.node.Kind == yaml.ScalarNode {
+				dest.Set(reflect.ValueOf(src.node.Value))
+			} else {
+				dest.Set(reflect.ValueOf(fmt.Sprintf("%v", reflectedSrc.Interface())))
+			}
+		}
+		return true, nil
+	}
+
+	// if we have a collection don't proceed to attempt to unmarshal direct
+	// from the yaml.Node ... collections are process per item, rather than
+	// as a whole.
+	if isCollection(dest) {
+		return false, errors.WithStack(
+			notAssignableError{
+				srcType:        reflectedSrc.Type(),
+				dstType:        dest.Type(),
+				sourceLocation: NewSource(m.sourceFile, WithLocation(coord)),
+			},
+		)
+	}
+
+	if !isSpecial(dest) {
+		if dest.CanAddr() {
+			meth := dest.Addr().MethodByName("UnmarshalYAML")
```

Comment:
Yeah, unfortunately there are two unmarshaller interfaces, only one of them is public, we want to allow for both since yaml.v3 does:
https://github.com/go-yaml/yaml/blob/v3.0.1/yaml.go#L34-L42

We could copy the interfaces locally to test agains as you suggest.